addons/payment_paypal/models/paypal_s2s.py

```diff
@@ -0,0 +1,231 @@
+        token_response = urllib2.urlopen(token_request)
+        token_result = json.loads(token_response.read())
+        token_response.close()
+        return token_result.get('access_token')
+
+
+class TxPaypalS2s(models.Model):
+    _inherit = 'payment.transaction'
+
+    # --------------------------------------------------
+    # SERVER2SERVER RELATED METHODS
+    # --------------------------------------------------
+
+    def _paypal_try_url(self, request, tries=3):
+        done, res = False, None
+        while (not done and tries):
+            try:
+                res = urllib2.urlopen(request)
+                done = True
+            except urllib2.HTTPError as e:
+                res = e.read()
+                e.close()
+                if tries and res and json.loads(res)['name'] == 'INTERNAL_SERVICE_ERROR':
+                    _logger.warning('Failed contacting Paypal, retrying (%s remaining)' % tries)
+            tries = tries - 1
+        if not res:
+            pass
+        result = json.loads(res.read())
+        res.close()
+        return result
+
+    def _paypal_s2s_send(self):
+        path = '/v1/payments/payment'
+        payment_url = self.env['payment.acquirer']._get_paypal_urls(self.acquirer_id.environment)['paypal_rest_url'] + path
+
+        amount_total = self.amount
+        amount = "%0.2f" % float(amount_total)
+
+        access_token = self.acquirer_id._paypal_s2s_get_access_token()
+        payment_headers = {
+            'Authorization': 'Bearer %s' % access_token,
+            'Content-Type': 'application/json'
+        }
+        data = {
+            'intent': 'sale',
+            'payer': {
+                'payment_method': 'credit_card',
+                'funding_instruments': [{
+                    'credit_card_token': {
+                        'credit_card_id': str(self.payment_method_id.acquirer_ref),
+                        'payer_id': str(self.payment_method_id.partner_id.id)
+                    }
+                }]
+            },
+            'transactions': [{
+                'amount': {
+                    'total': str(amount),
+                    'currency': str(self.currency_id.name),
+                },
+                'description': str(self.display_name)
+            }]
+        }
+        payment_result = dict()
+        try:
+            payment_request = urllib2.Request(payment_url,
+                                              json.dumps(data),
+                                              payment_headers)
+            payment_result = self._paypal_try_url(payment_request, tries=3)
+
+        except Exception, e:
+            _logger.exception('PayPal create payment request failed - %s' % e)
+            raise UserWarning(_('PayPal create payment request failed.'))
+        return payment_result
+
+    @api.multi
+    def paypal_s2s_do_transaction(self, **vals):
+        self.ensure_one()
+        payment_method = self.env['payment.method'].search([
```

Comment:
No, something's wrong here.

If a partner has multiple credit cards, only the first one will ever be debited.

The ´payment.transaction´ object should have the ´payment_method_id´ field set at creation, and you should use the many2one reference instead of this blind search.